lib/ansible/galaxy/collection/gpg.py

```diff
@@ -0,0 +1,298 @@
+# -*- coding: utf-8 -*-
+# Copyright: (c) 2022, Ansible Project
+# GNU General Public License v3.0+ (see COPYING or https://www.gnu.org/licenses/gpl-3.0.txt)
+"""Signature verification helpers."""
+
+from ansible.errors import AnsibleError
+from ansible.galaxy.user_agent import user_agent
+from ansible.module_utils.common.text.converters import to_text
+from ansible.module_utils.urls import open_url
+
+import os
+import subprocess
+import textwrap
+
+from dataclasses import dataclass, fields as dc_fields
+from urllib.error import HTTPError, URLError
+
+
+def run_gpg_verify(b_manifest_file, signature, keyring, display):
+    status_fd_read, status_fd_write = os.pipe()
+
+    cmd = [
+        'gpg',
+        f'--status-fd={status_fd_write}',
+        '--verify',
+        '--batch',
+        '--no-tty',
+        '--no-default-keyring',
+        f'--keyring={keyring}',
+        '-',
+        b_manifest_file,
+    ]
+    cmd_str = ' '.join(to_text(arg) for arg in cmd)
+    display.vvvv(f"Running command '{cmd}'")
+
+    try:
+        p = subprocess.Popen(
+            cmd,
+            stdin=subprocess.PIPE,
+            stdout=subprocess.PIPE,
+            stderr=subprocess.PIPE,
+            pass_fds=(status_fd_write,),
+        )
+    except (FileNotFoundError, subprocess.SubprocessError) as err:
+        raise AnsibleError(
+            f"Failed during GnuPG verification with command '{cmd_str}': {err}"
+        ) from err
+    else:
+        stdout, stderr = p.communicate(input=signature)
+    finally:
+        os.close(status_fd_write)
+
+    with os.fdopen(status_fd_read) as f:
+        stdout = f.read()
+        display.vvvv(f"{stdout} (exit code {p.returncode})")
+        return stdout, p.returncode
+
+
+def parse_gpg_errors(status_out):
+    for line in status_out.splitlines():
+        if not line:
+            continue
+        try:
+            _dummy, status, remainder = line.split(maxsplit=2)
+        except ValueError:
+            _dummy, status = line.split(maxsplit=1)
+            remainder = None
+        try:
+            cls = GPG_ERROR_MAP[status]
+        except KeyError:
+            continue
+
+        fields = [status]
+        if remainder:
+            fields.extend(
+                remainder.split(
+                    None,
+                    len(dc_fields(cls)) - 2
+                )
+            )
+
+        yield cls(*fields)
+
+
+def get_signature_from_source(source, display=None):
```

Comment:
```suggestion
def get_signature_from_source(source: Union[bytes, str], display: bool = None) -> str:
```
(it doesn't return bytes, does it?)

Reply:
Just returns str type. Fixed (now this is a method in ConcreteArtifactsManager).